packages/next-swc/crates/next-core/src/app_structure.rs

```diff
@@ -771,23 +771,53 @@ async fn directory_tree_to_entrypoints_internal(
                 components: components.without_leafs().cell(),
             }
             .cell();
+
             add_app_page(
                 app_dir,
                 &mut result,
                 "/not-found".to_string(),
                 "/not-found".to_string(),
-                tree,
+                dev_not_found_tree,
             )
             .await?;
             add_app_page(
                 app_dir,
                 &mut result,
                 "/_not-found".to_string(),
                 "/_not-found".to_string(),
-                tree,
+                dev_not_found_tree,
             )
             .await?;
         }
+    } else {
+        // Create default not-found page for production if there's no customized
+        // not-found
+        let prod_not_found_tree = LoaderTree {
+            segment: directory_name.to_string(),
+            parallel_routes: indexmap! {
+                "children".to_string() => LoaderTree {
+                    segment: "__PAGE__".to_string(),
+                    parallel_routes: IndexMap::new(),
+                    components: Components {
+                        page: Some(get_next_package(app_dir).join("dist/client/components/not-found-error.js".to_string())),
+                        ..Default::default()
+                    }
+                    .cell(),
+                }
+                .cell(),
+            },
+            components: components.without_leafs().cell(),
+        }
+        .cell();
+
+        add_app_page(
+            app_dir,
+            &mut result,
+            "/_not-found".to_string(),
+            "/_not-found".to_string(),
+            prod_not_found_tree,
+        )
+        .await?;
     }
```

Comment:
Can you comment why this needs to separate than the `path_prefix == "/"` branch above?

Reply:
@jridgewell  The condition was incorrect, I filed a fix for that here. The `else` should be attached to when `components.not_found` is not present

https://github.com/vercel/next.js/pull/54687